src/storage/src/hummock/store/version.rs

```diff
@@ -235,3 +260,369 @@ impl HummockReadVersion {
         self.staging.sst.clear();
     }
 }
+
+pub struct HummockReadSnapshot(
+    pub Vec<ImmutableMemtable>,
+    pub Vec<SstableInfo>,
+    pub CommittedVersion,
+);
+
+impl HummockReadSnapshot {
+    pub fn build_for_batch(
+        epoch: HummockEpoch, // for check
+        table_id: TableId,
+        key_range: &(Bound<Vec<u8>>, Bound<Vec<u8>>),
+        read_version_vec: Vec<Arc<RwLock<HummockReadVersion>>>,
+    ) -> StorageResult<Self> {
+        assert!(!read_version_vec.is_empty());
+        let read_version_guard_vec = read_version_vec
+            .iter()
+            .map(|read_version| read_version.read())
+            .collect_vec();
+        let mut imm_vec = Vec::default();
+        let mut sst_vec = Vec::default();
+        let mut lastst_committed_version =
+            read_version_guard_vec.get(0).unwrap().committed().clone();
+        let mut max_mce = 0;
+
+        // to get max_mce with lock_guard to avoid loosing committed_data since the read_version
+        // update is asynchronous
+        for read_version_guard in &read_version_guard_vec {
+            let committed_version = read_version_guard.committed.clone();
+            if committed_version.max_committed_epoch() > max_mce {
+                max_mce = committed_version.max_committed_epoch();
+                lastst_committed_version = committed_version;
+            }
+        }
+
+        let query_epoch = std::cmp::max(max_mce, epoch);
+        // prune imm and sst with max_mce
+        for read_version_guard in read_version_guard_vec {
+            let (imm_iter, sst_iter) =
+                read_version_guard
+                    .staging()
+                    .prune_overlap(query_epoch, table_id, key_range);
```

Comment:
Seems that `prune_overlap` cannot be used here, since it includes the epoch that `<= query_epoch`, while we expect the data with epoch `> max MCE`